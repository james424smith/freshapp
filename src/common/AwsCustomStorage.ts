import Keychain from "react-native-secure-storage";
const memoryKeyPrefix = "@MemoryStorage:";

interface MemoryData {
  [key: string]: string;
}
async function handleItemInKeychain(
  key: string,
  method: string,
  value?: string
) {
  try {
    if (method === "get") {
      return Keychain.getItem(key);
    } else if (method === "remove") {
      return Keychain.removeItem(key);
    } else {
      return Keychain.setItem(key, value);
    }
  } catch (error) {
    throw Error(error);
  }
}

export default class AmplifyCommunityAsyncStorage {
  static memoryData: MemoryData = {};

  static multiGet = async function (keys: string[], type: string) {
    return Promise.all(
      keys.map(async (key) => {
        await new Promise((res) => setTimeout(res, 100)); // sleep for 100ms
        return [
          key,
          await handleItemInKeychain(key, type).catch((error) =>
            console.log(`Error while trying ${type} for ${key}`, error)
          ),
        ];
      })
    );
  };

  static syncPromise: void | undefined;
  static setItem = async (key: string, value: string) => {
    try {
      await handleItemInKeychain(`${memoryKeyPrefix}${key}`, "set", value);
    } catch (error) {
      console.info(`Failed to set item ${key} in AsyncStorage: ${error}`);
    }
    AmplifyCommunityAsyncStorage.memoryData[key] = value;
    return value;
  };
  static getItem = (key: string) => {
    const value = AmplifyCommunityAsyncStorage.memoryData[key];
    return value === undefined ? null : value;
  };

  static removeItem = async (key: string) => {
    try {
      await handleItemInKeychain(`${memoryKeyPrefix}${key}`, "remove");
      delete AmplifyCommunityAsyncStorage.memoryData[key];
    } catch (error) {
      console.info(`Failed to remove item ${key} in AsyncStorage: ${error}`);
    }
  };
  static clear = async () => {
    const keys = await Keychain.getAllKeys();
    const memoryKeys = keys.filter((key: string) =>
      key.startsWith(memoryKeyPrefix)
    );
    await AmplifyCommunityAsyncStorage.multiGet(memoryKeys, "remove");
    AmplifyCommunityAsyncStorage.memoryData = {};
  };
  static createSyncPromise = async (_temp?: string) => {
    AmplifyCommunityAsyncStorage.memoryData = {};
    const keys = await Keychain.getAllKeys();
    const memoryKeys = keys.filter((key: string) =>
      key.startsWith(memoryKeyPrefix)
    );
    const pairs = await AmplifyCommunityAsyncStorage.multiGet(
      memoryKeys,
      "get"
    );
    pairs.forEach(([memoryKey, value]) => {
      if (value) {
        const key = (memoryKey as string).replace(memoryKeyPrefix, "");
        AmplifyCommunityAsyncStorage.memoryData[key] = value;
      }
    });
  };

  static sync = async () => {
    console.log(AmplifyCommunityAsyncStorage.syncPromise);
    if (!AmplifyCommunityAsyncStorage.syncPromise) {
      AmplifyCommunityAsyncStorage.syncPromise = await AmplifyCommunityAsyncStorage.createSyncPromise();
    }
    return AmplifyCommunityAsyncStorage.syncPromise;
  };
}
