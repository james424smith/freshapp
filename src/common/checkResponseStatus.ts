export const checkResponseStatus = (error: { status: number }) => {
  if (error.status === 404) {
    return [];
  } else {
    throw error;
  }
};
