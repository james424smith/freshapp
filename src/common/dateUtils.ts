import { DateTime, DurationObject } from "luxon";
import moment from "moment-timezone";

export type DateObject = {
  day: number;
  month: number;
  year: number;
};

export const getTimestampFromDate = (date?: string): number =>
  date
    ? DateTime.fromFormat(date, "dd-MM-yyyy", { zone: "UTC" }).toMillis()
    : 0;

export const timeDiffTextMillis = (fromDate?: number, toDate?: number) => {
  if (fromDate && toDate) {
    const luxonFrom = DateTime.fromMillis(fromDate, { zone: "UTC" });
    const luxonTo = DateTime.fromMillis(toDate, { zone: "UTC" });
    const diff: DurationObject = luxonTo
      .diff(luxonFrom, ["days", "hours", "minutes", "seconds"])
      .toObject();

    if (diff.days && diff.days > 0) {
      if (diff.days && diff.days < 7) {
        return `${Math.ceil(Number(diff.days))}d`;
      } else {
        return luxonFrom.toFormat("dd.MM.yy");
      }
    } else if (diff.hours && diff.hours > 0) {
      return `${Math.ceil(Number(diff.hours))}h`;
    } else if (diff.minutes && diff.minutes > 0) {
      return `${Math.ceil(Number(diff.minutes))}m`;
    } else {
      return `1m`;
    }
  }
  throw Error("fromDate and toDate must both be defined");
};

export const formatDate = (date?: Date, format = "dd.MM.yyyy") =>
  date ? DateTime.fromJSDate(date).toFormat(format) : "";

export const timeToNow = (fromDate?: number, toDate?: number) => {
  if (fromDate && toDate) {
    const luxonFrom = DateTime.fromMillis(fromDate, { zone: "UTC" });
    const luxonTo = DateTime.fromMillis(toDate, { zone: "UTC" });
    const diff: DurationObject = luxonTo
      .diff(luxonFrom, ["hours", "minutes"])
      .toObject();

    return diff.minutes === 0 ? `${diff.hours}h` : `<${(diff.hours || 0) + 1}h`;
  } else {
    throw Error("fromDate and toDate must both be defined");
  }
};
export const timestampToUtcString = (timestamp?: number) => {
  if (timestamp) {
    const date = DateTime.fromMillis(timestamp, { zone: "UTC" });
    return date.toFormat("dd.MM.yy HH:mm");
  } else {
    throw Error("timestamp must be defined");
  }
};

export const timestampToLocalString = (timestamp?: number) => {
  if (timestamp) {
    const date = DateTime.fromMillis(timestamp, { zone: "UTC" }).toLocal();
    return date.toFormat("dd.MM.yy    HH:mm");
  } else {
    throw Error("timestamp must be defined");
  }
};

export const compareToDateFromDateStartOfDay = (toDate: DateObject) => {
  const { year, month, day } = toDate;
  /**
   * Month starts from 00. so January is 00, February 01, etc.
   */
  if (year && month && day) {
    return (
      new Date().getTime() >=
      new Date(toDate.year, toDate.month - 1, toDate.day).getTime()
    );
  } else {
    throw Error("year, month, and date must all be defined");
  }
};

export const getTimeFromTimezone = (
  dateTime: Date,
  timeZone?: string,
  format = "dd.MM.yyyy HH:mm"
) => {
  if (timeZone) {
    return moment(dateTime, format).tz(timeZone).format(format);
  } else {
    return moment(dateTime, format).format(format);
  }
};
