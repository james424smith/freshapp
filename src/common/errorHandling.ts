export default (errorCode?: string) => {
  switch (errorCode) {
    case "NetworkError":
      return "You are offline. Please check your internet connection.";
    case "signUpOracle":
      return "Email address entered does not match our records. In order to proceed with registration please either enter a different email address or contact our Support Team (feedback.app@marlowgroup.com) in order to update your records. Please always include your Marlow ID when contacting our Support Team.";
    case "signUpAws":
      return "There is a problem. Seems you already started or completed the registration process. You need to go back to the main screen and login with your Marlow ID and password. If you did not create a personal password yet then please check your email account for the temp password.";
    case "UserNotFoundException":
      return "Please make sure your Employee Id and/or password are correct.";
    case "Login":
      return "There is a problem. Did you already create an account? If not, please select the 'Create New Account' option and then copy / paste the password from the email. If you already created an account please ensure you are entering the correct Marlow ID and personal password. If you don't remember your personal password please use the 'Forgot Password' option";
    case "ForgotPassword":
      return "A verification code has been sent to the registered email of the provided Employee ID. Please use this code to reset your password.";
    default:
      return "There is a problem. Did you already create an account? If not, please select the 'Create New Account' option and then copy / paste the password from the email. If you already created an account please ensure you are entering the correct Marlow ID and personal password. If you don't remember your personal password please use the 'Forgot Password' option";
  }
};
