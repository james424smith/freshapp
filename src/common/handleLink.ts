import { Linking } from "react-native";

export const handleLink = (
  redirectType: string,
  destination?: string
): Promise<void> | "" | undefined => {
  if (redirectType === "url") {
    return destination && Linking.openURL(destination);
  } else if (redirectType === "crewcompanion") {
    return destination && Linking.openURL(`crewcompanion://${destination}`);
  } else {
    return destination && Linking.openURL(`${redirectType}:${destination}`);
  }
};
