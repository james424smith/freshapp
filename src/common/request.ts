import config from "../constants/config";
import { RequestMethods } from "../interfaces";
import * as RequestHelpers from "./request";
import {
  assignmentsEndpoint,
  imprintEndpoint,
  seafarerEndpoint,
  signUpEndpoint,
  workingClothesEndpoint,
  newsEndpoint,
  termsAndConditionsEndpoint,
  privacyPolicyEndpoint,
  seafarerDocumentsEndpoint,
  seafarerDocumentsFileEndpoint,
  seaServiceEndpoint,
  notificationsEndpoint,
  notificationReadEndpoint,
  seafarerBadgesEndpoint,
  contactEndpoint,
  payslipEndpoint,
  allotmentsEndpoint,
  payslipDocumentsFileEndpoint,
  flightsEndpoint,
  validateWebsocketEndpoint,
  covidDocumentEndpoint,
  endpointOptions,
  checkVersionEndpoint,
} from "../constants/api_endpoints";
import assignmentApi from "../api/assignmentDetailsApi";
import imprintApi from "../api/imprintDetailsApi";
import seafarerApi from "../api/seafarerDetailsApi";
import { Auth } from "aws-amplify";
import newsApi from "../api/newsApi";
import termsAndConditionsApi from "../api/termsAndConditionsApi";
import workingClothesApi from "../api/workingClothesApi";
import privacyPolicyApi from "../api/privacyPolicyApi";
import seafarerDocumentsApi from "../api/seafarerDocumentsApi";
import seaServiceApi from "../api/seaServiceApi";
import notificationsApi from "../api/notificationsApi";
import { getAllotmentsDetails, getPayslipDetails } from "../api/salaryApi";
import seafarerBadgesApi from "../api/seafarerBadgesApi";
import contactApi from "../api/contactApi";
import flightsApi from "../api/flightDetailsApi";
import covidDocumentApi from "../api/covidDocumentApi";
import versionCheckApi from "../api/versionCheckApi";

type RequestParams = {
  requestMethod: RequestMethods;
  endpoint: endpointOptions;
  documentId?: number;
  documentCounter?: number;
  payload?: any;
  headers?: any;
  navigation?: any;
  vesselCode?: string;
  payslipId?: string;
};

export const getJwtToken = async () => {
  const session = await Auth.currentSession();
  return {
    Authorization: session.getAccessToken().getJwtToken(),
    seafarerId: session.getAccessToken().payload.username,
  };
};

const handleErrors = (response: any) => {
  if (!response.ok) {
    throw response;
  }
  return response;
};

export const chooseFetcher = async (request: {
  url: string;
  userData?: Record<string, string>;
  method: RequestMethods;
  headers?: Record<string, unknown>;
  body?: any;
}) => {
  return fetch(request.url, {
    body: request.body ? JSON.stringify(request.body) : undefined,
    method: request.method,
    headers: {
      ...(request.headers ?? {}),
      Authorization: request.userData ? request.userData.Authorization : "",
      "Content-Type": "application/json",
    },
  }).then(handleErrors);
};

const chooseDummyData = (requestParams: RequestParams) => {
  switch (requestParams.endpoint) {
    case "assignment":
      return assignmentApi().then((assignment) => assignment);
    case "imprint":
      return imprintApi().then((imprint) => imprint);
    case "news":
      return newsApi().then((news) => news);
    case "versionCheck":
      return versionCheckApi().then((versions) => versions);
    case "termsAndConditions":
      return termsAndConditionsApi().then(
        (termsAndConditions) => termsAndConditions
      );
    case "workingClothes":
      return workingClothesApi().then((workingClothes) => workingClothes);
    case "seafarerDocuments":
      return seafarerDocumentsApi().then(
        (seafarerDocuments) => seafarerDocuments
      );
    case "privacyPolicy":
      return privacyPolicyApi().then((privacyPolicy) => privacyPolicy);
    case "seaService":
      return seaServiceApi().then((seaService) => seaService);
    case "payslip":
      return getPayslipDetails().then((payslip) => payslip);
    case "allotments":
      return getAllotmentsDetails().then((allotments) => allotments);
    case "notifications":
      return notificationsApi().then((notifications) => notifications);
    case "seafarerBadges":
      return seafarerBadgesApi().then((seafarerBadges) => seafarerBadges);
    case "seafarerDocumentsFile":
      return Promise.resolve({
        json: () => Promise.resolve({ document: "12344566", fileType: "pdf" }),
      });
    case "payslipDocumentsFile":
      return Promise.resolve({
        json: () =>
          Promise.resolve({ payslip: "12344566", documentType: "pdf" }),
      });
    case "contact":
      return contactApi().then((contact) => contact);
    case "flights":
      return flightsApi().then((contact) => contact);
    case "covid":
      return covidDocumentApi().then((covidDocument) => covidDocument);
    case "validateWebsocket":
      return Promise.resolve(true);
    default:
      return seafarerApi().then((seafarer) => seafarer);
  }
};

export default async (requestParams: RequestParams): Promise<any> => {
  const { ENV, API_BASE_URL } = config();
  if (ENV === "DUMMY") {
    return chooseDummyData(requestParams);
  } else {
    switch (requestParams.endpoint) {
      case "assignment": {
        // call the function to get username and JWT Tokens
        const userData = await RequestHelpers.getJwtToken();
        const url = assignmentsEndpoint(API_BASE_URL, userData.seafarerId);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
          userData,
        });
      }
      case "imprint": {
        const url = imprintEndpoint(API_BASE_URL);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
        });
      }
      case "signup": {
        const url = signUpEndpoint(API_BASE_URL);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
          body: requestParams.payload,
        });
      }
      case "news": {
        const url = newsEndpoint(API_BASE_URL);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
        });
      }
      case "workingClothes": {
        // call the function to get username and JWT Tokens
        const userData = await RequestHelpers.getJwtToken();
        const url = workingClothesEndpoint(API_BASE_URL, userData.seafarerId);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
          userData,
        });
      }
      case "termsAndConditions": {
        const url = termsAndConditionsEndpoint(API_BASE_URL);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
        });
      }
      case "privacyPolicy": {
        const url = privacyPolicyEndpoint(API_BASE_URL);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
        });
      }
      case "seaService": {
        const userData = await RequestHelpers.getJwtToken();
        const url = seaServiceEndpoint(API_BASE_URL, userData.seafarerId);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
          userData,
        });
      }
      case "seafarerBadges": {
        const userData = await RequestHelpers.getJwtToken();
        const url = seafarerBadgesEndpoint(API_BASE_URL, userData.seafarerId);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
          userData,
        });
      }
      case "payslip": {
        const userData = await RequestHelpers.getJwtToken();
        const url = payslipEndpoint(API_BASE_URL, userData.seafarerId);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
          userData,
        });
      }
      case "payslipDocumentsFile": {
        const userData = await RequestHelpers.getJwtToken();
        const url = payslipDocumentsFileEndpoint(
          API_BASE_URL,
          userData.seafarerId,
          requestParams.vesselCode,
          requestParams.payslipId
        );
        return RequestHelpers.chooseFetcher({
          url: url,
          method: requestParams.requestMethod,
          userData,
        });
      }
      case "allotments": {
        const userData = await RequestHelpers.getJwtToken();
        const url = allotmentsEndpoint(API_BASE_URL, userData.seafarerId);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
          userData,
        });
      }
      case "notifications": {
        const userData = await RequestHelpers.getJwtToken();
        const url = notificationsEndpoint(API_BASE_URL, userData.seafarerId);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
          userData,
        });
      }
      case "notificationRead": {
        const userData = await RequestHelpers.getJwtToken();
        const url = notificationReadEndpoint(API_BASE_URL);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
          headers: {
            Accept: "application/json; charset=utf-8",
            "Access-Control-Allow-Origin": "*",
          },
          userData,
          body: {
            ...requestParams.payload,
            seafarerId: userData.seafarerId,
          },
        });
      }
      case "seafarerDocuments": {
        // call the function to get username and JWT Tokens
        const userData = await RequestHelpers.getJwtToken();
        const url = seafarerDocumentsEndpoint(
          API_BASE_URL,
          userData.seafarerId
        );
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
          userData,
        });
      }
      case "flights": {
        // call the function to get username and JWT Tokens
        const userData = await RequestHelpers.getJwtToken();
        const url = flightsEndpoint(API_BASE_URL, userData.seafarerId);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
          userData,
        });
      }
      case "seafarerDocumentsFile": {
        // call the function to get username and JWT Tokens
        const userData = await RequestHelpers.getJwtToken();
        const url = seafarerDocumentsFileEndpoint(
          API_BASE_URL,
          userData.seafarerId,
          requestParams.documentId,
          requestParams.documentCounter
        );
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
          userData,
        });
      }
      case "contact": {
        const userData = await RequestHelpers.getJwtToken();
        const url = contactEndpoint(API_BASE_URL, userData.seafarerId);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
          userData,
        });
      }
      case "covid": {
        const url = covidDocumentEndpoint(API_BASE_URL);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
        });
      }
      case "validateWebsocket": {
        const userData = await RequestHelpers.getJwtToken();
        const url = validateWebsocketEndpoint(
          API_BASE_URL,
          userData.seafarerId
        );
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
          userData,
        });
      }
      case "versionCheck": {
        const url = checkVersionEndpoint(API_BASE_URL);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
        });
      }
      default: {
        // call the function to get username and JWT Tokens
        const userData = await RequestHelpers.getJwtToken();
        const url = seafarerEndpoint(API_BASE_URL, userData.seafarerId);
        return RequestHelpers.chooseFetcher({
          url,
          method: requestParams.requestMethod,
          userData,
        });
      }
    }
  }
};
