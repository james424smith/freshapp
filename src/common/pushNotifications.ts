import { Auth } from "aws-amplify";
import * as RNFS from "react-native-fs";
import AsyncStorage from "@react-native-community/async-storage";
import { SIGN_IN_ROUTE } from "../constants/routes";
import { AppState, Platform, Vibration, PushNotification } from "react-native";
import PushNotificationIOS from "@react-native-community/push-notification-ios";
import { ANDROID, IOS } from "../constants/Platforms";
import {
  ASSIGNMENT,
  CHAT,
  FLIGHT,
  GENERAL,
  NEWSLETTER,
  PAYSLIP,
  PERSONAL,
  DELETE_USER,
  LOGOUT,
} from "../constants/notificationsConstants";
import * as PushNotificationsHelper from "./pushNotifications";
import PushNotificationAmplify from "@aws-amplify/pushnotification";

import {
  SET_NEW_CHAT_NOTIFICATION,
  PushNotificationType,
  InitialisedSagaCall,
  WebsocketEvent,
} from "../interfaces";
import { handleLink } from "./handleLink";

const errorMessage = "error while redirecting to screen";

export const handleSignOut = async (shouldBeGlobal: boolean) => {
  const currentEmployeeId = await AsyncStorage.getItem("employeeId");
  const shouldSaveEmployeeId = await AsyncStorage.getItem(
    "shouldSaveEmployeeId"
  );

  await Auth.signOut({ global: shouldBeGlobal });
  const documentPath = `${RNFS.DocumentDirectoryPath}/document`;
  const documentFolder = await RNFS.exists(documentPath);
  documentFolder && (await RNFS.unlink(documentPath));
  const payslipPath = `${RNFS.DocumentDirectoryPath}/payslip`;
  const payslipFolder = await RNFS.exists(payslipPath);
  payslipFolder && (await RNFS.unlink(payslipPath));
  const asyncStorageKeys = await AsyncStorage.getAllKeys();
  if (asyncStorageKeys.length > 0) {
    Promise.all(
      asyncStorageKeys.map(async (key: string) => {
        if (key !== "colorScheme" && key !== "SIGN_IN_PATH") {
          await AsyncStorage.removeItem(key);
        }
      })
    );
  }

  currentEmployeeId &&
    (await AsyncStorage.setItem("employeeId", currentEmployeeId));
  shouldSaveEmployeeId &&
    (await AsyncStorage.setItem("shouldSaveEmployeeId", shouldSaveEmployeeId));
};

type Props = {
  getNotificationsDetails: () => InitialisedSagaCall;
  dispatch: (arg?: unknown) => void;
  navigation: any;
  setWebsocketData: (event: WebsocketEvent) => InitialisedSagaCall;
};

export const extractNotificationData = (notification: PushNotificationType) => {
  if (Platform.OS === ANDROID) {
    return notification.data
      ? JSON.parse(notification.data["pinpoint.jsonBody"])
      : JSON.parse(notification["pinpoint.jsonBody"]);
  } else {
    return notification._data.data.jsonBody;
  }
};

export const extractNotificationBody = (notification: PushNotificationType) => {
  if (notification.body) {
    return notification.body;
  } else {
    return notification._alert?.body;
  }
};

export const handlePushNotificationRedirection = async ({
  notificationCategory,
}: {
  notificationCategory: string;
}) => {
  switch (notificationCategory) {
    case DELETE_USER:
      await PushNotificationsHelper.handleSignOut(true);
      break;
    case LOGOUT:
      await PushNotificationsHelper.handleSignOut(false);
      break;
    case ASSIGNMENT:
      await handleLink("crewcompanion", "assignment");
      break;
    case PAYSLIP:
      await handleLink("crewcompanion", "salaries");
      break;
    case FLIGHT:
      await handleLink("crewcompanion", "flights");
      break;
    case CHAT:
      await handleLink("crewcompanion", "chat/history");
      break;
    case PERSONAL:
      await handleLink("crewcompanion", "documents");
      break;
    case NEWSLETTER:
      await handleLink("crewcompanion", "news");
      break;
    case GENERAL:
      await handleLink("crewcompanion", "notifications");
      break;
    default:
      await handleLink("crewcompanion", "notifications");
      break;
  }
};

export default async () => {
  // // required on iOS only (see fetchCompletionHandler docs: https://facebook.github.io/react-native/docs/pushnotificationios.html)
  if (Platform.OS === IOS) {
    PushNotificationIOS.getInitialNotification().then(
      (notification: PushNotification | null) => {
        if (notification !== null) {
          const notificationData = PushNotificationsHelper.extractNotificationData(
            (notification as unknown) as PushNotificationType
          );
          PushNotificationsHelper.handlePushNotificationRedirection(
            notificationData
          )
            .then(() => {})
            .catch((error) => console.log(errorMessage, error));
          // required on iOS only (see fetchCompletionHandler docs: https://facebook.github.io/react-native/docs/pushnotificationios.html)
          notification.finish(PushNotificationIOS.FetchResult.NoData);
        }
      }
    );
  }
  // get the notification data when notification is received
  PushNotificationAmplify.onNotification(
    async (notification: PushNotification | null) => {
      Vibration.vibrate([0, 500, 100, 200]);
      const notificationData = PushNotificationsHelper.extractNotificationData(
        (notification as unknown) as PushNotificationType
      );
      if (Platform.OS === ANDROID) {
        if (notificationData.notificationCategory === "DeleteUser") {
          PushNotificationsHelper.handlePushNotificationRedirection(
            notificationData
          ).catch((error) => console.log(errorMessage, error));
        }
      } else {
        if (AppState.currentState === "background" && notification) {
          PushNotificationsHelper.handlePushNotificationRedirection(
            notificationData
          ).catch((error) => console.log(errorMessage, error));
          // required on iOS only (see fetchCompletionHandler docs: https://facebook.github.io/react-native/docs/pushnotificationios.html)
          notification.finish(PushNotificationIOS.FetchResult.NoData);
        }
      }
    }
  );

  // get the notification data when notification is opened
  PushNotificationAmplify.onNotificationOpened(
    async (notification: PushNotification | null) => {
      const notificationData = PushNotificationsHelper.extractNotificationData(
        (notification as unknown) as PushNotificationType
      );
      await PushNotificationsHelper.handlePushNotificationRedirection(
        notificationData
      ).catch((error) => console.log(errorMessage, error));
    }
  );
};

export const getNotificationOnForeground = async (props: Props) =>
  PushNotificationAmplify.onNotification(
    async (notification: PushNotification | null) => {
      Vibration.vibrate([0, 500, 100]);
      console.log("in app notification", notification);
      props.getNotificationsDetails();
      const notificationData = PushNotificationsHelper.extractNotificationData(
        (notification as unknown) as PushNotificationType
      );
      if (notificationData.notificationCategory === DELETE_USER) {
        await PushNotificationsHelper.handleSignOut(true).then(() => {
          props.navigation.navigate("Auth", { screen: SIGN_IN_ROUTE });
        });
      }
      if (notificationData.notificationCategory === LOGOUT) {
        await PushNotificationsHelper.handleSignOut(false).then(() =>
          props.navigation.navigate("Auth", { screen: SIGN_IN_ROUTE })
        );
      }
      if (notificationData.notificationCategory === CHAT) {
        const event = {
          messageType: SET_NEW_CHAT_NOTIFICATION,
          payload: {
            ...notificationData,
            notificationBody: PushNotificationsHelper.extractNotificationBody(
              (notification as unknown) as PushNotificationType
            ),
          },
        };
        props.dispatch(props.setWebsocketData(event as WebsocketEvent));
      }
    }
  );
