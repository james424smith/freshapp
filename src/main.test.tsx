import React from "react";
import { cleanup, render, waitFor } from "@testing-library/react-native";
import * as RequestHelpers from "./common/request";
import DeviceInfo from "react-native-device-info";
import { Alert, Platform } from "react-native";
import { Analytics } from "aws-amplify";
import MainScreen from "./index";

describe("MainScreen snapshot test", () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  afterAll((done) => {
    done();
  });

  it("should render MainScreen ", async () => {
    Analytics.updateEndpoint = jest.fn().mockResolvedValue({});
    const { toJSON } = render(<MainScreen />);
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it("should throw alert when platform is ios and version does not match", async () => {
    Platform.OS = "ios";
    DeviceInfo.getVersion = jest.fn().mockReturnValue("1.3.0");
    const spy = jest
      .spyOn(RequestHelpers, "chooseFetcher")
      .mockResolvedValue({ data: { ios: "1.3.1" } });
    const spy2 = jest.spyOn(Alert, "alert");

    render(<MainScreen />);

    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledTimes(1);
    });
    spy.mockRestore();
    spy2.mockRestore();
  });
  it("should throw alert when platform is android and version does not match", async () => {
    Platform.OS = "android";
    DeviceInfo.getVersion = jest.fn().mockReturnValue("1.3.0");
    const spy = jest
      .spyOn(RequestHelpers, "chooseFetcher")
      .mockResolvedValue({ data: { android: "1.3.1" } });
    const spy2 = jest.spyOn(Alert, "alert");

    render(<MainScreen />);

    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledTimes(1);
    });
    spy.mockRestore();
    spy2.mockRestore();
  });
  it("should throw alert when platform is android and version does match", async () => {
    Platform.OS = "android";
    DeviceInfo.getVersion = jest.fn().mockReturnValue("1.3.1");
    const spy = jest
      .spyOn(RequestHelpers, "chooseFetcher")
      .mockResolvedValue({ data: { android: "1.3.1" } });
    const spy2 = jest.spyOn(Alert, "alert");

    render(<MainScreen />);

    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledTimes(0);
    });
    spy.mockRestore();
    spy2.mockRestore();
  });
  it("should throw alert when platform is ios and version does match", async () => {
    Platform.OS = "ios";
    DeviceInfo.getVersion = jest.fn().mockReturnValue("1.3.1");
    const spy = jest
      .spyOn(RequestHelpers, "chooseFetcher")
      .mockResolvedValue({ data: { android: "1.3.1" } });
    const spy2 = jest.spyOn(Alert, "alert");

    render(<MainScreen />);

    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledTimes(0);
    });
    spy.mockRestore();
    spy2.mockRestore();
  });
});
