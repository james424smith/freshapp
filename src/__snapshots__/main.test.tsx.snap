// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainScreen snapshot test should render MainScreen  1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <RNCSafeAreaProvider
        onInsetsChange={[Function]}
        style={
          Array [
            Object {
              "flex": 1,
            },
            undefined,
          ]
        }
      >
        <View
          onLayout={[Function]}
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "opacity": 1,
                }
              }
            />
            <View
              accessibilityElementsHidden={false}
              closing={false}
              gestureVelocityImpact={0.3}
              importantForAccessibility="auto"
              onClose={[Function]}
              onGestureBegin={[Function]}
              onGestureCanceled={[Function]}
              onGestureEnd={[Function]}
              onOpen={[Function]}
              onTransition={[Function]}
              pointerEvents="box-none"
              style={
                Array [
                  Object {
                    "overflow": undefined,
                  },
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                ]
              }
              transitionSpec={
                Object {
                  "close": Object {
                    "animation": "spring",
                    "config": Object {
                      "damping": 500,
                      "mass": 3,
                      "overshootClamping": true,
                      "restDisplacementThreshold": 10,
                      "restSpeedThreshold": 10,
                      "stiffness": 1000,
                    },
                  },
                  "open": Object {
                    "animation": "spring",
                    "config": Object {
                      "damping": 500,
                      "mass": 3,
                      "overshootClamping": true,
                      "restDisplacementThreshold": 10,
                      "restSpeedThreshold": 10,
                      "stiffness": 1000,
                    },
                  },
                }
              }
            >
              <View
                needsOffscreenAlphaCompositing={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "flex": 1,
                    "marginTop": 0,
                  }
                }
              >
                <View
                  collapsable={false}
                  forwardedRef={[Function]}
                  onGestureHandlerEvent={[Function]}
                  onGestureHandlerStateChange={[Function]}
                  style={
                    Object {
                      "flex": 1,
                      "transform": Array [
                        Object {
                          "translateX": 0,
                        },
                        Object {
                          "translateX": 0,
                        },
                      ],
                    }
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "flex": 1,
                          "overflow": "hidden",
                        },
                        Array [
                          Object {
                            "backgroundColor": "rgb(242, 242, 242)",
                          },
                          Object {
                            "borderBottomWidth": 0,
                            "elevation": 0,
                            "shadowColor": "transparent",
                          },
                        ],
                      ]
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                          "flexDirection": "column-reverse",
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "flex": 1,
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "backgroundColor": "rgb(242, 242, 242)",
                              "flex": 1,
                            }
                          }
                        >
                          <View
                            style={
                              Object {
                                "alignItems": "center",
                                "flex": 1,
                                "justifyContent": "center",
                                "marginBottom": 112.5,
                              }
                            }
                          >
                            <Image
                              source={Object {}}
                              style={
                                Object {
                                  "height": 400,
                                  "width": 400,
                                }
                              }
                            />
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          pointerEvents="box-none"
          style={
            Object {
              "height": 64,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": 1,
            }
          }
        />
      </RNCSafeAreaProvider>
    </View>
  </View>
</View>
`;
